Compatibility with Embroider

I'm not sure if there's much to do here besides some documentation around Embroider but thought I'd raise it anyway.

### Problem

It appears things are stabilising around Embroider because since ember-cli 3.25.0 [all newly-created addons will have Embroider testing baked into ember-try](https://github.com/ember-cli/ember-cli/pull/9436) by default. I've been experimenting with Embroider and various addons and it appears that `ember-cli-string-helpers` is incompatible with an Embroider app when you use the `only` or `except` configuration.

If you try and run a build using the following configuration it fails with the 

```js
'ember-cli-string-helpers': {
  only: ['dasherize', 'underscore'],
  except: ['titleize', 'capitalize'],
},
```

```
ERROR Summary:

  - broccoliBuilderErrorStack: ModuleNotFoundError: Module not found: Error: Can't resolve '../node_modules/ember-cli-string-helpers/helpers/capitalize' in '$TMPDIR/embroider/f661dd/helpers'
    at /Users/pat/hacking/string-helpers-embroider/node_modules/webpack/lib/Compilation.js:925:10
    at /Users/pat/hacking/string-helpers-embroider/node_modules/webpack/lib/NormalModuleFactory.js:401:22
    at /Users/pat/hacking/string-helpers-embroider/node_modules/webpack/lib/NormalModuleFactory.js:130:21
    at /Users/pat/hacking/string-helpers-embroider/node_modules/webpack/lib/NormalModuleFactory.js:224:22
    at /Users/pat/hacking/string-helpers-embroider/node_modules/neo-async/async.js:2830:7
    at /Users/pat/hacking/string-helpers-embroider/node_modules/neo-async/async.js:6877:13
    at /Users/pat/hacking/string-helpers-embroider/node_modules/webpack/lib/NormalModuleFactory.js:214:25
    at /Users/pat/hacking/string-helpers-embroider/node_modules/enhanced-resolve/lib/Resolver.js:213:14
    at /Users/pat/hacking/string-helpers-embroider/node_modules/enhanced-resolve/lib/Resolver.js:285:5
    at eval (eval at create (/Users/pat/hacking/string-helpers-embroider/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
```

I have a reproduction in this repo [here](https://github.com/patocallaghan/string-helpers-embroider). You can see the configuration used in https://github.com/patocallaghan/string-helpers-embroider/commit/39003b207455995d48f3d65a31239e0c77b4f33a

### Embroider support
It's worth calling out that the use case for using the `only` and `except` configuration will no longer be required by Embroider "Optimized" mode, i.e. tree-shaking enabled, as that will only load the helpers in your build that you specifically use.

I guess the problem with this addon is if folks are using Embroider "Safe" mode which aims to offer full backwards compatibility with the existing Ember CLI pipeline.

### Fix?
I'm not sure if it's worth fixing this as personally I'm just going to remove the `only` and `except` config while we are experimenting with Embroider "Safe" mode. Maybe all that is required is some documentation on Embroider compatibility in the README. With Embroider "Optimized" mode the use case for `only` and `except` is no longer needed.

--- a/.travis.yml
+++ b/.travis.yml
@@ -44,6 +44,8 @@ jobs:
     - env: EMBER_TRY_SCENARIO=ember-beta
     - env: EMBER_TRY_SCENARIO=ember-canary
     - env: EMBER_TRY_SCENARIO=ember-default-with-jquery
+    - env: EMBER_TRY_SCENARIO=embroider-safe
+    - env: EMBER_TRY_SCENARIO=embroider-optimized
 
 before_install:
   - npm config set spin false

--- a/config/ember-try.js
+++ b/config/ember-try.js
@@ -1,6 +1,7 @@
 'use strict';
 
 const getChannelURL = require('ember-source-channel-url');
+const { embroiderSafe, embroiderOptimized } = require('@embroider/test-setup');
 
 module.exports = function() {
   return Promise.all([
@@ -76,7 +77,9 @@ module.exports = function() {
               '@ember/jquery': '^0.5.1'
             }
           }
-        }
+        },
+        embroiderSafe(),
+        embroiderOptimized()
       ]
     };
   });

--- a/ember-cli-build.js
+++ b/ember-cli-build.js
@@ -1,6 +1,7 @@
 'use strict';
 
 const EmberAddon = require('ember-cli/lib/broccoli/ember-addon');
+const { maybeEmbroider } = require('@embroider/test-setup');
 
 module.exports = function(defaults) {
   let app = new EmberAddon(defaults, {
@@ -14,5 +15,5 @@ module.exports = function(defaults) {
     behave. You most likely want to be modifying `./index.js` or app's build file
   */
 
-  return app.toTree();
+  return maybeEmbroider(app);
 };

--- a/index.js
+++ b/index.js
@@ -4,6 +4,7 @@ var Funnel = require('broccoli-funnel');
 var path = require('path');
 var intersection = require('./lib/intersection');
 var difference = require('./lib/difference');
+var StripBadReexports = require('./lib/strip-bad-reexports');
 
 module.exports = {
   name: require('./package').name,
@@ -16,6 +17,8 @@ module.exports = {
       app = app.app;
     }
 
+    this.app = app;
+
     var addonOptions = (this.parent && this.parent.options) || (this.app && this.app.options) || {};
     var addonConfig = addonOptions[this.name] || {};
 
@@ -23,12 +26,15 @@ module.exports = {
     this.blacklist = this.generateBlacklist(addonConfig);
   },
 
-  treeForAddon: function() {
-    // see: https://github.com/ember-cli/ember-cli/issues/4463
-    var tree = this._super.treeForAddon.apply(this, arguments);
-    var moduleRegexp = '(^modules\/)?' + this.name + '\/helpers\/';
+  treeForAddon: function(tree) {
+    tree = this.filterHelpers(tree, new RegExp(/^helpers\//, 'i'));
+    tree = new StripBadReexports(tree, [`index.js`]);
+    return this._super.treeForAddon.call(this, tree);
+  },
 
-    return this.filterHelpers(tree, new RegExp(moduleRegexp, 'i'));
+  treeForApp: function(tree) {
+    tree = this.filterHelpers(tree, new RegExp(/^helpers\//, 'i'));
+    return this._super.treeForApp.call(this, tree);
   },
 
   filterHelpers: function(tree, regex) {
@@ -37,8 +43,8 @@ module.exports = {
     var _this = this;
 
     // exit early if no opts defined
-    if (whitelist.length === 0 && blacklist.length === 0) {
-      return new Funnel(tree);
+    if ((!whitelist || whitelist.length === 0) && (!blacklist || blacklist.length === 0)) {
+      return tree;
     }
 
     return new Funnel(tree, {

--- a/lib/strip-bad-reexports.js
+++ b/lib/strip-bad-reexports.js
@@ -1,0 +1,53 @@
+/* eslint-env node */
+const resolve = require('resolve');
+const { transform } = require('@babel/core');
+const Funnel = require('broccoli-funnel');
+const { writeFileSync, readFileSync, existsSync, unlinkSync } = require('fs');
+const { join } = require('path');
+
+
+module.exports = class StripBadReexports extends Funnel {
+  constructor(tree, files) {
+    super(tree);
+    this.filesToStrip = files;
+  }
+  build() {
+    super.build();
+    for (let file of this.filesToStrip) {
+      let srcFile = join(this.inputPaths[0], file);
+      if (!existsSync(srcFile)) {
+        continue;
+      }
+      let src = readFileSync(srcFile, 'utf8');
+      let plugins = [[stripBadReexportsTransform, { resolveBase: this.outputPath }]];
+      let destFile = join(this.outputPath, file);
+      unlinkSync(destFile);
+      writeFileSync(destFile, transform(src, { plugins }).code);
+    }
+  }
+  shouldLinkRoots() {
+    // We want to force funnel to copy things rather than just linking the whole
+    // directory, because we're planning to mutate it.
+    return false;
+  }
+}
+
+function stripBadReexportsTransform() {
+  return {
+    visitor: {
+      ExportNamedDeclaration(path, state) {
+        if (
+          path.node.source &&
+          path.node.source.type === 'StringLiteral'
+        ) {
+          try {
+            resolve.sync(path.node.source.value, { basedir: state.opts.resolveBase });
+          } catch (err) {
+            path.remove();
+          }
+        }
+      },
+    },
+  };
+}
+

--- a/package.json
+++ b/package.json
@@ -61,8 +61,10 @@
     "test": "tests"
   },
   "dependencies": {
+    "@babel/core": "^7.13.10",
     "broccoli-funnel": "^3.0.3",
-    "ember-cli-babel": "^7.7.3"
+    "ember-cli-babel": "^7.7.3",
+    "resolve": "^1.20.0"
   },
   "engines": {
     "node": "10.* || >=12.*"

--- a/package.json
+++ b/package.json
@@ -15,6 +15,7 @@
   "homepage": "https://github.com/romulomachado/ember-cli-string-helpers",
   "devDependencies": {
     "@ember/optional-features": "^2.0.0",
+    "@embroider/test-setup": "^0.37.0",
     "broccoli-asset-rev": "^3.0.0",
     "chai": "^4.2.0",
     "ember-ajax": "^5.0.0",

